### PR TITLE
[REF] fields: use date and datetime values in fields

### PR DESCRIPTION
**Pull request by kujiu**
Create new classes to return date and datetime-like objects accepting string behaviors. This permits to be compatible with existing code and to manipulate date and datetime without need of manual conversion. Each datetime has an explicit timezone, UTC by default, to fix all issues about misconvertion when timezone is unknown.

Exemples of operations:
```python
record.start_datetime -= datetime.timedelta(days=2)
record.start_datetime.replace('2017', '2018')
record.start_datetime.replace(year=2018)
record.start_datetime - '2015-10-26 16:29:00'
```

New classes are based on [pendulum](https://pendulum.eustace.io/) to have native support of correct timezones and a lot of great features. Date and datetime are automatically converted in good format for JSON exchange. The datetime lib is added to context for safe eval and server actions.

This cleans all modules and converts them to the new behavior.

### CODE REVIEW

**Comment by sylvain-garancher** [comment]
:+1: 
Great !

**Comment by nhomar** [comment]
Finally... REALLY NICE Let's work on this ASAP to have it stable for v12

File: addons/account/wizard/account_report_aged_partner_balance.py [hunk 1]
```diff
@@ -27,14 +25,14 @@ def _print_report(self, data):
...
-        start = datetime.strptime(data['form']['date_from'], "%Y-%m-%d")
+        start = date.from_string(data['form']['date_from'])
 
         for i in range(5)[::-1]:
             stop = start - relativedelta(days=period_length - 1)
             res[str(i)] = {
-                'name': (i!=0 and (str((5-(i+1)) * period_length +1) + '-' + str((5-i) * period_length)) or ('+'+str(4 * period_length))),
-                'stop': start.strftime('%Y-%m-%d'),
-                'start': (i!=0 and stop.strftime('%Y-%m-%d') or False),
+                'name': (i!=0 and (str((5-(i+1)) * period_length) + '-' + str((5-i) * period_length)) or ('+'+str(4 * period_length))),
```
Comment (rco-odoo) [inline addons/account/wizard/account_report_aged_partner_balance.py]: Why is the "+1" gone on this line?

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -590,12 +594,11 @@ def check_full_reconcile(self):
         total_debit = 0
         total_credit = 0
         total_amount_currency = 0
-        maxdate = '0000-00-00'
         to_balance = {}
         for aml in amls:
             total_debit += aml.debit
             total_credit += aml.credit
-            maxdate = max(aml.date, maxdate)
+            maxdate = max(aml.date, datelib.max)
```
Comment (rco-odoo) [inline addons/account/models/account_move.py]: should be `datelib.min`

**Comment by kujiu** [comment]
This pull request is closed. We will do a new branch without datetime wrapper, without pendulum, without timezone management and with a minimum of changes.

**Comment by christophlsa** [comment]
Can you explain why you now go another way? I understand that this was probably more R&D but would be nice to get some background because the examples in the ticket description were interesting. :)

**Comment by kujiu** [comment]
We have decided to change the scope of this task, and to reimplement with these constraints:
- no dependency to an external library;
- compatibility with standard naive datetime;
- avoid to bother developers by fixing a timezone (timezone issues are the responsability of the developer and not of the ORM);
- no wrapper for datetime;
- minimal changes.

In the new way, all datetimes will be naive in UTC by convention.

**Pull request closed without merging**